Seoungdeok jeon/fire bar

cjhj0320 opened the pull request:
# Pull Request Description

Please fill out the form below. For the checkbox, use [x] to have a checkmark. For example:

- [x] With a checkmark.
- [ ] Without a checkmark.

This should not affect the way whether your pull request is accepted or not. Think of it like a survey and documentation, so the reviewers and future developers can follow your PR easier.

1. Have you included *.gitignore* file? Or manually remove config and environment files (E.g: *node_modules* folder, Eclipse IDE files) from this submission? The answer should be **Yes** in this case.

   - [x] Yes
   - [ ] No

2. Did you fully complete the assigned task(s)?

   - [x] Fully completed
   - [ ] Partially completed

3. Did you introduce a new non-required feature in this PR? This feature is not an assigned task, but it is the one that you feel will make the game better. For example, reproduce a bug used by speed runner in an original version.

   - [ ] Yes
   - [x] No

4. Are you aware of any bug caused by your code that have not been fixed in this PR?

   - [ ] Yes
   - [x] No

5. Are you aware of any **potential bug**? You don't have to fix this, it is just a way to document. You may not face the bug during the development process, but theoretically it can happens in edge cases. For example, memory leak or a player falls out of map.

   - [ ] Yes
   - [x] No

## Unimplented task(s)
None.

## Related issue(s)
- Fire Bar

## Added
In sprites folder:
- fireBar_fire.png image

In bricks.js:
- FireBar class
- Fire class

In main.js:
- "ASSET_MANAGER.queueDownload("./sprites/firebar_fire.png");"

## Changed and Enhanced
update() method in mario.js:
- if (entity instanceof Fire) { that.die(); }

loadLevel() method in sceneManager.js:
- let firebar = new FireBar(this.game, 10* PARAMS.BLOCKWIDTH, 10* PARAMS.BLOCKWIDTH);
  this.game.addEntity(firebar);
(I added the above lines of code just for demonstrating purpose. Please remove the above line when actually implementing the levels.)

## Fixed
None.

## Bugs
1. Not-fixed bugs (only answer when you chose Yes in question 4):
- None known.

2. Potential bugs (only answer when you chose Yes in question 5):
- None known.

## Contributors and code author(s)
- Seoungdeok Jeon

## Final check

Have you fully filled out this document, and add a label **PR: NEED REVIEW** to this pull request? Answer *Yes* (as it should be) means you are good to go! Thanks for the contribution.

- [x] Yes
- [ ] No

BenjaminDeJager commented on bricks.js:
```diff
@@ -208,4 +208,75 @@ class SideTube {
+        Object.assign(this, { game, x, y });
+        let brick = new Brick(this.game, this.x, this.y, 3, "None");
+        this.game.addEntity(brick);
+
```
![image](https://user-images.githubusercontent.com/61642215/108308012-d689fb00-7163-11eb-9bc8-da89d2ccfa93.png)
https://www.mariowiki.com/Fire_Bar#Super_Mario_Bros.
I see that your hard-coding the length/number of fireballs. but I was not sure if its always 6 fireballs and I I THINK that its implying that they are variable so you might need to add a 4th parameter or something for the number of fireballs? I'm honestly not sure tho.

cjhj0320 replied:
I think every firebar has 6 fireballs in Super Mario Bros.

hunghvu replied:
![image](https://user-images.githubusercontent.com/62678592/108567319-e832e600-72bc-11eb-823f-7d5282a8e000.png)

BenjaminDeJager reviewed (approved):
I am noting that the firebar enemy might have variable length so this enemy might need a 4th parameter to allow for the scenemanager to specify the length. so perhaps something like: 
       "let firebar = new FireBar(this.game, 10* PARAMS.BLOCKWIDTH, 10* PARAMS.BLOCKWIDTH, 6); "
when the level has 
       " firebars: [{ x: 20, y: 10, 6 }]"

Also, I see you have it defined in brick.js instead of enemys.js and I think I why, although it might end up moved to the enemys.js file

Neither of these are reasons why this can't just be straight merged with main so I'm approving this pr

BenjaminDeJager commented on bricks.js:
```diff
@@ -208,4 +208,75 @@ class SideTube {
+    };
+}
+
+class Fire {
```
do note the existence of the "fireball" class added by avatarabo for their "firemario" task (a flower that lets mario shoot fireballs). no conflict here fortunately, but I wanted to mention this when I noticed

hunghvu reviewed (changes requested):
Based on #8, you completed No.1, 2, 4, and 5. Variable length is also requested in the PR (No.3), which is not completed. You don't have to complete all bullet points so that's fine. Besides, the bounding box of firebar is a bit too small to me. Admittedly I'm not familiar with Mario, but it seems like the fire bar intersects with Mario too much. This is technically not a big issue though.

However, I will request change to this PR due to a _potential **game breaking bug**_ where in some scenarios, when Mario is killed by the fire bar, he is teleported to a different location on a map.

The _bounding box issue_ (0:00) and _a teleportation bug_ I mentioned (0:41) can be seen in the following video.

https://user-images.githubusercontent.com/62678592/108567734-a3f41580-72bd-11eb-96d4-7098f225d179.mp4

cjhj0320 commented:
I just updated the code according to Hung's and Benjamin's suggestions.
Here are the changes that I made:
1. Added a 4th parameter for the number of fireballs.
2. Increased the size of the bounding box of fire.
3. Change the name of the "Fire" class to "FireBar_Fire".
4. Moved FireBar and FireBar_Fire classes to enemies.js file.

Also, I think the game breaking bug that Hung mentioned is a bug related to die() method in Mario class. I did not change anything in that method, and I'm just calling it in the update() method. 
![image](https://user-images.githubusercontent.com/63382084/108609149-e7fc1e80-7380-11eb-98fb-a0934c046d67.png)
Should I modify the die() method that is in the Mario class? I'm just worried to fix it because I think that is not my assigned task.

hunghvu reviewed (dismissed):
Looks good to me, and you're right, in **Mario.js: die(), line this.velocity.y = -640**, causes the issue. That line is to make Mario move up when he's dead. What I notice is, this bug will always happen when that line exists, no matter what the value is. If we comment out the line, the bug disappears, but Mario will not move up when being dead anymore. I also have been looking at this bug in the last 30 minutes but still cannot find out the root cause.

**I will approve the PR, and you don't need to fix the bug as you are not responsible for that death mechanism.** I suppose everyone will have the same kind of issue.

BenjaminDeJager commented on enemies.js:
```diff
@@ -238,4 +238,78 @@ class PirahnaPlant {
+    };
+}
+
+class FireBar_Fire {
```
I see u renamed the class to remove potential conflict, good idea even if not required.

BenjaminDeJager commented on enemies.js:
```diff
@@ -238,4 +238,78 @@ class PirahnaPlant {
+};
+
+class FireBar {
+    constructor(game, x, y, numOfFires) {
```
right, you agreed to assume the number of fireballs was variable.

BenjaminDeJager commented on enemies.js:
```diff
@@ -238,4 +238,78 @@ class PirahnaPlant {
+        }
+    };
+
+    drawMinimap(ctx, mmX, mmY) {
```
the firebar_fires probably shouldn't show up to avoid cluttering the mini map...but the fire-bar itself might benefit from being added....maybe just draw a 1-pixel orange line that rotates around the block'd origin to show current rotation?

alternatively 2 orange pixels with 1 being the origin and the 2'd used to reflect the outer-most ball's current location?

BenjaminDeJager commented on enemies.js:
```diff
@@ -238,4 +238,78 @@ class PirahnaPlant {
+        if(this.angle >= 360){
+            this.angle = 0;
+        }
+        console.log(this.numOfFires);
```
this needs to go before final implementation irrc. but even if our project does not have the "no console outputs" policy I hope you know that its considered bad practice to leave console-outputs in most situations....especially in a method that gets called as much as update().

I strongly recommend that you, at the least, wrap such statements in a "if debug" check.

BenjaminDeJager commented on enemies.js:
```diff
@@ -238,4 +238,78 @@ class PirahnaPlant {
+    update() {
+        this.angle += this.game.clockTick;
+        if(this.angle >= 360){
+            this.angle = 0;
```
you shouldn't set the angle to 0 on line 267 as that will make 367 degree's jump to 0 (along with every value after 360.) instead you should use mod 360 so that values outside of [0, 360) are mapped to it correctly.

the mod- operator in JS is indeed the same one as in java and python so "367 % 360" = "7"

http://www.java2s.com/Tutorial/JavaScript/0040__Operators/Modulus.htm

in particular, I suggest "this.angle = this.angle % 360"

even better however would be if line 265 just directly set this.angle to the modded value. 
this.angle = (this.angle + this.game.clockTick) mod 360

BenjaminDeJager reviewed (approved):
despite the issues I saw in main.js, I will approve this as they will not actually break anything and thus not required to be fixed (I doubt the player will notice the firebar teleporting around, at most it will be pixel's at a time I believe).

.
